internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,382 @@
+	m.mu.RLock()
+	tlsCert := m.tlsCert
+	idCert := m.idCert
+	m.mu.RUnlock()
+	if idCert != nil && isCertActive(idCert) && tlsCert != nil && isCertActive(tlsCert.Leaf) {
+		return tlsCert, nil
+	}
+
+	// Slow path: generate a new in-memory identity signed TLS certificate.
+	//
+	// If we got here then it means we need to generate a new in-memory TLS
+	// keypair, and potentially an identity certificate (only the first time
+	// or when it expires, or when the key changed).
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if err := m.createDir(); err != nil {
+		return nil, fmt.Errorf("cannot create TLS directory: %w", err)
+	}
+	if err := m.getIDCert(); err != nil {
+		return nil, fmt.Errorf("cannot get identity certificate: %w", err)
+	}
+	if err = m.createTLSCert(); err != nil {
+		return nil, fmt.Errorf("cannot create TLS certificate: %w", err)
+	}
+	return m.tlsCert, nil
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *TLSManager) Ensure() error {
+	return nil
+}
+
+func (m *TLSManager) createTLSCert() error {
+	_, privateKey, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return err
+	}
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return err
+	}
+	template := x509.Certificate{
+		SerialNumber:          serialNumber,
+		NotBefore:             systemTime(),
+		NotAfter:              systemTime().Add(tlsCertValidity),
+		KeyUsage:              x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+	// DER encoded bytes
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, m.idCert, privateKey.Public(), m.signer)
+	if err != nil {
+		return err
+	}
+	// Load a clean X509 certificate from the DER encoded bytes.
+	certificate, err := x509.ParseCertificate(derBytes)
+	if err != nil {
+		return err
+	}
+	m.tlsCert = &tls.Certificate{
+		PrivateKey: privateKey,
+		// Leaf first, then CA cert.
+		Certificate: [][]byte{certificate.Raw, m.idCert.Raw},
+		Leaf:        certificate,
+	}
+	return nil
+}
+
+// createDir verifies the directory layout and permissions, and creates
+// the leaf element of the TLS driectory if it does not yet exist.
+func (m *TLSManager) createDir() error {
+	exists, err := pathExists(m.tlsDir)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return expectPermission(m.tlsDir, 0o700)
+	}
+
+	// Create the leaf directory node with 0o700 permissions.
+	return os.Mkdir(m.tlsDir, 0o700)
+}
+
+// getIDCert verifies that the identity certificate exists, and that the
+// certificate matches the supplied identity key signer. If the identity key
+// has changed, or the certificate has expired, the identity certificate is
+// replaced with a new identity certificate signed with the identity key.
+func (m *TLSManager) getIDCert() error {
+	idPath := filepath.Join(m.tlsDir, idCertFile)
+	exists, err := pathExists(idPath)
```

Comment:
Do we need to call this if `m.idCert != nil`?

Reply:
I've reshuffled it a little bit, and I think its looks better now. The path is now only checked for the case where no certificate was previously loaded. Resolving.